repo2docker/contentproviders/ckan.py

```diff
@@ -0,0 +1,101 @@
+import re
+from datetime import datetime, timedelta, timezone
+from os import path
+from urllib.parse import urlparse
+
+from requests import Session
+
+from .. import __version__
+from .base import ContentProvider
+
+
+class CKAN(ContentProvider):
+    """Provide contents of a remote CKAN dataset."""
+
+    def __init__(self):
+        super().__init__()
+        self.session = Session()
+        self.session.headers.update(
+            {
+                "user-agent": f"repo2docker {__version__}",
+            }
+        )
+
+    def _fetch_version(self, api_url):
+        """Fetch dataset modified date and convert to epoch.
+        Borrowed from the Hydroshare provider.
+        """
+        package_show_url = f"{api_url}package_show?id={self.dataset_id}"
+        resp = self.urlopen(package_show_url).json()
+        date = resp["result"]["metadata_modified"]
+        parsed_date = datetime.strptime(date, "%Y-%m-%dT%H:%M:%S.%f")
+        epoch = parsed_date.replace(tzinfo=timezone(timedelta(0))).timestamp()
+        # truncate the timestamp
+        return str(int(epoch))
+
+    def _request(self, url, **kwargs):
+        return self.session.get(url, **kwargs)
+
+    urlopen = _request
+    url_regex = r"/dataset/[a-z0-9_\\-]*$"
+
+    def detect(self, source, ref=None, extra_args=None):
+        """Trigger this provider for things that resolve to a CKAN dataset."""
+        parsed_url = urlparse(source)
+        if not parsed_url.netloc:
+            return None
+
+        api_url = parsed_url._replace(
+            path=re.sub(self.url_regex, "/api/3/action/", parsed_url.path)
```

Comment:
Instead of using a regex here, can we write out the splitting logic more directly? Something like (haven't tested this, but you get the gist):

```python
url_parts = parsed_url.path.split('/')
if url_parts[-2] == 'dataset':
   dataset_id = url_parts[-1]
else:
   return None
api_url_path = f'/api/3/action/{dataset_id}'
api_url = parsed_url.replace(path=api_url_path)
```

Reply:
This also lets us re-use `dataset_id` down below

Reply:
Thanks. I have made changes according to your suggestion.